src/backend/optimizer/path/allpaths.c

```diff
@@ -3004,24 +3005,43 @@ recurse_push_qual(Node *setOp, Query *topquery,
 }
 
 static bool
-is_query_contain_limit_groupby(Query *parse)
+check_query_for_possible_motion(Query *parse)
 {
+	ListCell   *lc;
+
+	// LIMIT, DISTINCT, GROUP BY, aggregate functions or recursive CTEs
+	// can cause motions in a plan, so return true if they are presented
 	if (parse->limitCount || parse->limitOffset ||
-		parse->groupClause || parse->distinctClause)
+		parse->groupClause || parse->distinctClause ||
+		parse->hasAggs || parse->hasRecursive)
 		return true;
 
+	// INTERSECT, EXCEPT or UNION without ALL set operations also can
+	// cause motions in a plan, so return true for them as well
 	if (parse->setOperations)
 	{
-		SetOperationStmt *sop_stmt = (SetOperationStmt *) (parse->setOperations);
-		RangeTblRef   *larg = (RangeTblRef *) sop_stmt->larg;
-		RangeTblRef   *rarg = (RangeTblRef *) sop_stmt->rarg;
-		RangeTblEntry *lrte = list_nth(parse->rtable, larg->rtindex-1);
-		RangeTblEntry *rrte = list_nth(parse->rtable, rarg->rtindex-1);
+		SetOperationStmt *stmt = (SetOperationStmt *) parse->setOperations;
+
+		if (!stmt->all ||
+			stmt->op == SETOP_INTERSECT ||
+			stmt->op == SETOP_EXCEPT)
+			return true;
+	}
+
+	foreach(lc, parse->cteList)
+	{
+		CommonTableExpr *cte = (CommonTableExpr *) lfirst(lc);
+
+		if (check_query_for_possible_motion((Query *) cte->ctequery))
+			return true;
+	}
+
+	foreach(lc, parse->rtable)
+	{
+		RangeTblEntry *rte = (RangeTblEntry *) lfirst(lc);
 
-		if ((lrte->rtekind == RTE_SUBQUERY &&
-			 is_query_contain_limit_groupby(lrte->subquery)) ||
-			(rrte->rtekind == RTE_SUBQUERY &&
-			 is_query_contain_limit_groupby(rrte->subquery)))
+		if (rte->rtekind == RTE_SUBQUERY &&
+			check_query_for_possible_motion(rte->subquery))
```

Comment:
How about small optimization?
```suggestion
		if (rte->rtekind == RTE_SUBQUERY)
			return check_query_for_possible_motion(rte->subquery);
```